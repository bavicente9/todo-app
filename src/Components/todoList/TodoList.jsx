
import { useDispatch, useSelector } from 'react-redux';
import { selectTheme } from '../../features/theme/themeSlice';
import { removeTodosCompleteds, selectCounterActive, selectEntities, selectFilter } from '../../features/todosList/todosSlice';
import FilterButtons from '../filterButtons/filterButtons';
import TodoItem from '../todoItem/TodoItem';
import './TodoList.scss'


const createTodosItems = (entities, filter) => {
    const todosItems = entities.map(item => {
        switch (filter) {
            case 'all':
                return (
                    <TodoItem active={item.active} id={item.id} text={item.text} key={item.id} />
                )

                break;
            case 'active':
                if (item.active) {
                    return (
                        <TodoItem active={item.active} id={item.id} text={item.text} key={item.id} />
                    )
                }
                break;


            case 'completed':
                if (!item.active) {
                    return (
                        <TodoItem active={item.active} id={item.id} text={item.text} key={item.id} />
                    )
                }
                break;
            default:
                return (
                    <TodoItem active={item.active} id={item.id} text={item.text} key={item.id} />
                )
                break;
        }
    })
     
    return (
        <>{todosItems}</>
    )
}


const TodoList = () => {

    const dispatch = useDispatch()
    const entities = useSelector(selectEntities)
    const theme = useSelector(selectTheme)
    let counterActiveTodos = useSelector(selectCounterActive)
    const filter = useSelector(selectFilter)

    const handleClearCompletedTodos = (e) => {
        e.preventDefault()
        dispatch(removeTodosCompleteds())
    }

    return (
        <>
            <div className={`todoList_container todoList_container--${theme}`}>
                <ul className='todoList_list'>
                    {createTodosItems(entities, filter)}
                </ul>

                <div className='footer'>
                    <p className='counter'>{counterActiveTodos} items left</p>

                    <FilterButtons className='filterButtons--desktop' />

                    <button onClick={e => handleClearCompletedTodos(e)} className='button_clear'>Clear completed</button>

                </div>


            </div>
            <FilterButtons className='filterButtons--mobile' />
        </>
    )
}

export default TodoList;