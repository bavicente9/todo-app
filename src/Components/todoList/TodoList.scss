@import '../../Utils/style/globalVariables.scss';
@import '../../Utils/style/checkboxTodo.scss';


.todoList_container{
    margin: 15px auto;
    min-width: $globalContainer-minWidth;
    max-width: $globalContainer-maxWidth;
    border-radius: 5px;
    
    .todoList_list{
        list-style: none;
        padding: 0;
        margin: 0;
        
        .todoList-item{ //TODO item
            display: grid;
            grid-template-columns: 40px auto;
            grid-template-rows: 1fr;
            align-items: center;
            justify-items: start;
            
            border-bottom: 1px solid $DarkGrayishBlue;
            p{
                text-align: start;
                margin:  15px 10px;
                overflow: scroll;
                color: $lightGrayishBlue;
            }
            
        }
        .item-completed{
            text-decoration: line-through;
        }
        
        //FINISH TODO ITEM
        
    }
    
    .footer{
        margin: 0 20px;
        display: flex;
        justify-content: space-between;
        align-items: center;
        
        .button_clear{
            background: none;
            border: none;
            cursor: pointer;
            color: inherit;
            &:hover{
                color: $lightGrayishBlueHover;
            }

        }

    }
    
    
}


.todoList_container--dark {
    background-color: $veryDarkGrayishBlue;
    color: $lightGrayishBlue;
}

.todoList_container--light {
    background-color: $veryLightGray;
    color: $veryDarkBlue;
    
}



@media (max-width:$globalMedia-maxWidth) {
    
    .todoList_container {
        min-width: $globalMobileContainer-minWidth;
        max-width: $globalMobileContainer-maxWidth;
        
    }
    
    
    
}