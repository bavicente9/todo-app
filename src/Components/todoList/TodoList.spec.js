import componentWithMockStore from "../../Utils/test/componentWithMockStore";
import renderer from 'react-test-renderer';
import TodoList from "./TodoList";




describe('TodoList', () => {
    let store
    let component;
    

    beforeEach(() => {
        //render a component with a mock Store
        ({ component, store } = componentWithMockStore(<TodoList />))
    });


    it('match with snapshot', () => {
        expect(component.toJSON()).toMatchSnapshot();
    })



    it('correctly call of clear Completed dispatch', () => {
        renderer.act(() => {
            const ev = {
                preventDefault: jest.fn()
            }
            component.root.findByProps({className:'button_clear'}).props.onClick(ev)
        }
        )
        expect(store.dispatch).toHaveBeenCalledTimes(1);
    })
    
})