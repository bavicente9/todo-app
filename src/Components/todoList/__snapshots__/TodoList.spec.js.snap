// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList match with snapshot 1`] = `
Array [
  <div
    className="todoList_container todoList_container--dark"
  >
    <ul
      className="todoList_list"
    >
      <li
        active="true"
        className="todoList-item item-active todoList-item--dark"
      >
        <input
          aria-label="completed checkbox"
          checked={false}
          className="input_checkBox"
          onChange={[Function]}
          type="checkbox"
        />
        <p>
          text example
        </p>
        <button
          aria-label="remove todo"
          className="todo_button--remove"
          onClick={[Function]}
        />
      </li>
      <li
        active="false"
        className="todoList-item item-completed todoList-item--dark"
      >
        <input
          aria-label="completed checkbox"
          checked={true}
          className="input_checkBox"
          onChange={[Function]}
          type="checkbox"
        />
        <p>
          text example number 2
        </p>
        <button
          aria-label="remove todo"
          className="todo_button--remove"
          onClick={[Function]}
        />
      </li>
    </ul>
    <div
      className="footer"
    >
      <p
        className="counter"
      >
        2
         items left
      </p>
      <div
        className="filterButtons_container filterButtons_container--dark filterButtons--desktop"
      >
        <button
          className="filter_button "
          onClick={[Function]}
          value="all"
        >
          All
        </button>
        <button
          className="filter_button  "
          onClick={[Function]}
          value="active"
        >
          Active
        </button>
        <button
          className="filter_button "
          onClick={[Function]}
          value="completed"
        >
          Completed
        </button>
      </div>
      <button
        className="button_clear"
        onClick={[Function]}
      >
        Clear completed
      </button>
    </div>
  </div>,
  <div
    className="filterButtons_container filterButtons_container--dark filterButtons--mobile"
  >
    <button
      className="filter_button "
      onClick={[Function]}
      value="all"
    >
      All
    </button>
    <button
      className="filter_button  "
      onClick={[Function]}
      value="active"
    >
      Active
    </button>
    <button
      className="filter_button "
      onClick={[Function]}
      value="completed"
    >
      Completed
    </button>
  </div>,
]
`;
